gradio/components/audio.py

```diff
@@ -102,24 +108,26 @@ def __init__(
             show_download_button: If True, will show a download button in the corner of the component for saving audio. If False, icon does not appear.
             show_share_button: If True, will show a share icon in the corner of the component that allows user to share outputs to Hugging Face Spaces Discussions. If False, icon does not appear. If set to None (default behavior), then the icon appears if this Gradio app is launched on Spaces, but not otherwise.
             show_edit_button: If True, will show an edit icon in the corner of the component that allows user to edit the audio. If False, icon does not appear. Default is True.
+            min_length: The minimum length of audio (in seconds) that the user can pass into the prediction function. If None, there is no minimum length.
+            max_length: The maximum length of audio (in seconds) that the user can pass into the prediction function. If None, there is no maximum length.
+            waveform_options: A dictionary of options for the waveform display. Options include: waveform_color (str), waveform_progress_color (str), show_controls (bool), skip_length (int). Default is None, which uses the default values for these options.
         """
-        valid_sources = ["upload", "microphone"]
-        source = source if source else ("microphone" if streaming else "upload")
-        if source not in valid_sources:
-            raise ValueError(
-                f"Invalid value for parameter `source`: {source}. Please choose from one of: {valid_sources}"
-            )
-        self.source = source
+        sources = (
+            list(sources)
+            if sources is not None
+            else (["microphone"] if streaming else ["microphone", "upload"])
+        )
```

Comment:
nit: handle the str case as well for better DX
```suggestion
        if sources is None:
            sources = ["microphone"] if streaming else ["microphone", "upload"]
        if isinstance(sources, str):
            sources = [sources]
        elif isinstance(sources, list):
            pass
        else:
            raise ValueError('`sources` must a list consisting of elements in ["microphone", "upload"]')
```